### PR TITLE
Adding test for StringBuilder.Clear() making sure Capacity wont grow unboundedly

### LINKED ISSUE #27625: System.Text.StringBuilder.Insert uses a lot of memory in certain scenarios
Ported from: https://developercommunity.visualstudio.com/content/problem/63747/systemtextstringbuilderinsert.html

Method System.Text.StringBuilder.Insert in a couple with System.Text.StringBuilder.Clear in some cases leads to very dramatic performance degradation and very high memory consumption. For the following test the StringBuilder class is absolutely unsuilable. But this case is only a simple try to reuse object in order to reduce the loading of GC. But the test shows that the result is absolutely opposite to expected.  

Test case:
        [TestMethod]
        public void TestStringBuilderInsert()
        {
            const int iterationGranularity = 10000;
            var buffer = new StringBuilder();
            var s = new string(' ', 10);
            Console.WriteLine($"{buffer.Length}->{buffer.Capacity}");
            var sw = Stopwatch.StartNew();
            for (int k = 0; k < 10; k++)
            {
                for (int i = 1; i <= iterationGranularity; i++)
                {
                    buffer.Clear();
                    buffer.Append(s);
                    buffer.Append(s);
                    buffer.Append(s);
                    buffer.Insert(0, s);
                    buffer.Insert(0, s);
                }
                Console.WriteLine($"{sw.ElapsedMilliseconds,5}: {buffer.Length}->{buffer.Capacity}");
            }
            sw.Stop();
        }

Test result:
0->16
   71: 50->100042
  251: 50->200042
  548: 50->300042
 1456: 50->400042
 2807: 50->500042
 4311: 50->600042
 6091: 50->700042
 8007: 50->800042
 9999: 50->900042
12203: 50->1000042

So, the question: If I never put into builder more than 50 characters at the iteration why the Capacity is over 1e6? 
Moreover if you use memory profiler, you will understand that the memory usage is much worse that it seems at first glance.
Screenshot of profiler I will attach. Do you think that allocation 100 Gb of memory is enought? I operated only with 10 Mb...

Name	Inclusive Allocations	Exclusive Allocations	Inclusive Bytes	Exclusive Bytes	Inclusive Allocations %
System.Char[]	200 018	                    200 018	               100 015 401 486	100 015 401 486	66,63

### PR DESCRIPTION
**Pull request by maryamariyan**
Fixes #27625

Tests for PR dotnet/coreclr#16926

### CODE REVIEW

File: src/System.Runtime/tests/System/Text/StringBuilderTests.cs [hunk 1]
```diff
@@ -979,6 +979,35 @@ public static void Clear()
...
+        {
+            var builder = new StringBuilder();
+            builder.Clear();
+            Assert.NotEqual(0, builder.Capacity);
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(10000)]
+        public static void Clear_AppendAndInsertBeforeClearManyTimes_CapacityStaysWithinRange(int times)
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/Text/StringBuilderTests.cs]: You will need to disable this for .NET Framework

Comment (danmosemsft) [reply src/System.Runtime/tests/System/Text/StringBuilderTests.cs]: With a note of the issue # fixing it, and that on NETFX you end up with 100042

**Comment by danmosemsft** [comment]
Please include link to CoreCLR fix in top comment

File: src/System.Runtime/tests/System/Text/StringBuilderTests.cs (hunk 1 above)
Comment (AlexGhiondea) [reply src/System.Runtime/tests/System/Text/StringBuilderTests.cs]: Suggestion: I would also add cases for the other possible outcomes of Clear (where we don't end up with ~1.2*Length of the original string)

**Comment by danmosemsft** [comment]
Consider checking there is a test for this not throwing.
```
var sb = new StringBuilder(14, 14);
sb.Append("Hello world!!!");
sb.Clear();
```

File: src/System.Runtime/tests/System/Text/StringBuilderTests.cs [hunk 2]
```diff
@@ -979,6 +979,58 @@ public static void Clear()
...
+        [Fact]
+        public static void Clear_Empty_CapacityNotZero()
+        {
+            var builder = new StringBuilder();
+            builder.Clear();
+            Assert.NotEqual(0, builder.Capacity);
+        }
+
+        [Theory]
+        [InlineData(14)]
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/Text/StringBuilderTests.cs]: Why use a Theory with only one value? That would normally be a fact, or you could add more data.

**Comment by maryamariyan** [comment]
@danmosemsft the two new tests below cover use case of having StringBuilder full at MaxCapacity: 

- ```Clear_AtMaxCapacity_SingleChunk_CapacityStaysUnchanged```
- ```Clear_AtMaxCapacity_SingleChunk_CapacityStaysUnchanged```

**Pull request merged**